sdk/src/transaction/sanitized.rs

```diff
@@ -0,0 +1,243 @@
+                    _ => false,
+                },
+            )
+            .filter(|ix| {
+                matches!(
+                    limited_deserialize(&ix.data),
+                    Ok(SystemInstruction::AdvanceNonceAccount)
+                )
+            })
+            .and_then(|ix| {
+                ix.accounts.get(0).and_then(|idx| {
+                    let idx = *idx as usize;
+                    self.message.get_account_key(idx)
+                })
+            })
+    }
+
+    /// Return the serialized message data to sign.
+    fn message_data(&self) -> Vec<u8> {
+        match &self.message {
+            SanitizedMessage::Legacy(message) => message.serialize(),
+            SanitizedMessage::V0(mapped_msg) => mapped_msg.message.serialize(),
+        }
+    }
+
+    /// Verify the length of signatures matches the value in the message header
+    pub fn verify_signatures_len(&self) -> bool {
+        self.signatures.len() == self.message.header().num_required_signatures as usize
+    }
+
+    /// Verify the transaction signatures
+    pub fn verify(&self) -> Result<()> {
+        let message_bytes = self.message_data();
+        if self
+            .signatures
+            .iter()
+            .zip(self.message.account_keys_iter())
+            .map(|(signature, pubkey)| signature.verify(pubkey.as_ref(), &message_bytes))
+            .any(|verified| !verified)
+        {
+            Err(TransactionError::SignatureFailure)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Verify the encoded secp256k1 signatures in this transaction
+    pub fn verify_precompiles(&self, libsecp256k1_0_5_upgrade_enabled: bool) -> Result<()> {
+        for (program_id, instruction) in self.message.program_instructions_iter() {
+            if secp256k1_program::check_id(program_id) {
+                let instruction_datas: Vec<_> = self
+                    .message
+                    .instructions()
+                    .iter()
+                    .map(|instruction| instruction.data.as_ref())
+                    .collect();
+                let data = &instruction.data;
+                let e = verify_eth_addresses(
+                    data,
+                    &instruction_datas,
+                    libsecp256k1_0_5_upgrade_enabled,
+                );
+                e.map_err(|_| TransactionError::InvalidAccountIndex)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Calculate the total fees for a transaction given a fee calculator
+    pub fn calculate_fee(&self, fee_calculator: &FeeCalculator) -> u64 {
+        let mut num_secp256k1_signatures: u64 = 0;
+        for (program_id, instruction) in self.message.program_instructions_iter() {
+            if secp256k1_program::check_id(program_id) && !instruction.data.is_empty() {
+                num_secp256k1_signatures =
+                    num_secp256k1_signatures.saturating_add(instruction.data[0] as u64);
```

Comment:
wdyt about replacing the `is_empty()`/`[0]` with `.get(0).and_then(...)` for a bit more maintenance resilience?

Reply:
Sure thing, replaced